openstack/identity/v3/extensions/oauth1/requests.go

```diff
@@ -0,0 +1,490 @@
+	_, r.Err = client.Get(userAccessTokenURL(client, userID, id), &r.Body, nil)
+	return
+}
+
+// RevokeAccessToken revokes an OAuth1 access token.
+func RevokeAccessToken(client *gophercloud.ServiceClient, userID string, id string) (r RevokeAccessTokenResult) {
+	_, r.Err = client.Delete(userAccessTokenURL(client, userID, id), nil)
+	return
+}
+
+// ListAccessTokens enumerates authorized access tokens.
+func ListAccessTokens(client *gophercloud.ServiceClient, userID string) pagination.Pager {
+	url := userAccessTokensURL(client, userID)
+	return pagination.NewPager(client, url, func(r pagination.PageResult) pagination.Page {
+		return AccessTokensPage{pagination.LinkedPageBase{PageResult: r}}
+	})
+}
+
+// ListAccessTokenRoles enumerates authorized access token roles.
+func ListAccessTokenRoles(client *gophercloud.ServiceClient, userID string, id string) pagination.Pager {
+	url := userAccessTokenRolesURL(client, userID, id)
+	return pagination.NewPager(client, url, func(r pagination.PageResult) pagination.Page {
+		return AccessTokenRolesPage{pagination.LinkedPageBase{PageResult: r}}
+	})
+}
+
+// GetAccessTokenRole retrieves details on a single OAuth1 access token role by
+// an ID.
+func GetAccessTokenRole(client *gophercloud.ServiceClient, userID string, id string, roleID string) (r GetAccessTokenRoleResult) {
+	_, r.Err = client.Get(userAccessTokenRoleURL(client, userID, id, roleID), &r.Body, nil)
+	return
+}
+
+// The following are small helper functions used to help build the signature.
+
+// buildAuthJSON builds a map, used to be passed to a Keystone V3 "/auth/token"
+// location.
+func buildAuthJSON() map[string]interface{} {
+	var req struct {
+		Auth struct {
+			Identity struct {
+				Methods []string `json:"methods"`
+				OAuth1  struct{} `json:"oauth1"`
+			} `json:"identity"`
+		} `json:"auth"`
+	}
+	req.Auth.Identity.Methods = []string{"oauth1"}
+	b, _ := gophercloud.BuildRequestBody(req, "")
+
+	return b
+}
+
+// tokenParams builds a URLEncoded parameters string.
+func tokenParams(query url.Values, timestamp *time.Time, nonce string, callback string) string {
+	if timestamp != nil {
+		// use provided timestamp
+		query.Set("oauth_timestamp", strconv.FormatInt(timestamp.Unix(), 10))
+	} else {
+		// use current timestamp
+		query.Set("oauth_timestamp", strconv.FormatInt(time.Now().UTC().Unix(), 10))
+	}
+
+	if nonce == "" {
+		// when nonce is not set, generate a random one
+		query.Set("oauth_nonce", strconv.FormatInt(rand.Int63(), 10)+query.Get("oauth_timestamp"))
+	}
+
+	if callback != "" {
+		query.Set("oauth_callback", callback)
+	}
+	query.Set("oauth_version", "1.0")
+
+	// sorted by key
+	return query.Encode()
+}
+
+// stringToSign builds a string to be signed.
+func stringToSign(method string, u string, params string) []byte {
+	parsedURL, _ := url.Parse(u)
+	if strings.HasPrefix(parsedURL.Host, "127.0.0.1") {
+		// Workaround for unit tests with dynamic ports
```

Comment:
I was able to get around this by doing the following:

```go
func TestAuthenticate(t *testing.T) {
  l, _ := net.Listen("tcp", "127.0.0.1:33199")
  th.Server = httptest.NewUnstartedServer(th.Mux)
  th.Server.Listener = l
  th.Server.Start()

  //th.SetupHTTP()
  defer th.TeardownHTTP()
  HandleAuthenticate(t)
...
```

However, the test is still failing, but I'm not sure why. 

I'd definitely like to see this conditional removed somehow, though.

Reply:
I created an extra testhelper function to listen the static port. Additionally I noticed that using "https://identity:443/" URL containing the default port is [not supported by oauth](https://oauth.net/core/1.0a/#rfc.section.9.1.2), therefore I modified the logic to normalize URLs with default port.